helper/ssh/key_pair.go

```diff
@@ -0,0 +1,268 @@
+// See the 'const' block for details.
+type KeyPairType string
+
+func (o KeyPairType) String() string {
+	return string(o)
+}
+
+// CreateKeyPairConfig describes how an SSH key pair should be created.
+type CreateKeyPairConfig struct {
+	// Type describes the key pair's type.
+	Type KeyPairType
+
+	// Bits represents the key pair's bits of entropy. E.g., 4096 for
+	// a 4096 bit RSA key pair, or 521 for a ECDSA key pair with a
+	// 521-bit curve.
+	Bits int
+
+	// Name is the resulting key pair's name. This is used to identify
+	// the key pair in the SSH server's 'authorized_keys'.
+	Name string
+}
+
+// FromPrivateKeyConfig describes how an SSH key pair should be loaded from an
+// existing private key.
+type FromPrivateKeyConfig struct {
+	// RawPrivateKeyPemBlock is the raw private key that the key pair
+	// should be loaded from.
+	RawPrivateKeyPemBlock []byte
+
+	// Name is the resulting key pair's name. This is used to identify
+	// the key pair in the SSH server's 'authorized_keys'.
+	Name string
+}
+
+// KeyPair represents an SSH key pair.
+// TODO: Maybe a field for a description? Maybe save the type?
+type KeyPair struct {
+	// PrivateKeyPemBlock represents the key pair's private key in
+	// ASN.1 Distinguished Encoding Rules (DER) format in a
+	// Privacy-Enhanced Mail (PEM) block.
+	PrivateKeyPemBlock []byte
+
+	// PublicKeyAuthorizedKeysLine represents the key pair's public key
+	// as a line in OpenSSH authorized_keys.
+	PublicKeyAuthorizedKeysLine []byte
+
+	// Name is the key pair's name. This is used to identify
+	// the key pair in the SSH server's 'authorized_keys'.
+	Name string
+}
+
+// KeyPairFromPrivateKey returns a KeyPair loaded from an existing private key.
+//
+// Supported key pair types include:
+// 	- DSA
+// 	- ECDSA
+// 	- ED25519
+// 	- RSA
+func KeyPairFromPrivateKey(config FromPrivateKeyConfig) (KeyPair, error) {
+	privateKey, err := gossh.ParseRawPrivateKey(config.RawPrivateKeyPemBlock)
+	if err != nil {
+		return KeyPair{}, err
+	}
+
+	switch pk := privateKey.(type) {
+	case *rsa.PrivateKey:
+		publicKey, err := gossh.NewPublicKey(&pk.PublicKey)
+		if err != nil {
+			return KeyPair{}, err
+		}
+		return KeyPair{
+			PrivateKeyPemBlock:          config.RawPrivateKeyPemBlock,
+			PublicKeyAuthorizedKeysLine: authorizedKeysLine(publicKey, config.Name),
+		}, nil
+	case *ecdsa.PrivateKey:
+		publicKey, err := gossh.NewPublicKey(&pk.PublicKey)
+		if err != nil {
+			return KeyPair{}, err
+		}
+		return KeyPair{
+			PrivateKeyPemBlock:          config.RawPrivateKeyPemBlock,
+			PublicKeyAuthorizedKeysLine: authorizedKeysLine(publicKey, config.Name),
+		}, nil
+	case *dsa.PrivateKey:
+		publicKey, err := gossh.NewPublicKey(&pk.PublicKey)
+		if err != nil {
+			return KeyPair{}, err
+		}
+		return KeyPair{
+			PrivateKeyPemBlock:          config.RawPrivateKeyPemBlock,
+			PublicKeyAuthorizedKeysLine: authorizedKeysLine(publicKey, config.Name),
+		}, nil
+	case *ed25519.PrivateKey:
```

Comment:
`*rsa.PrivateKey` and `*ecdsa.PrivateKey` also have a `Public() crypto.PublicKey` func. That seems to be the [`Signer`](https://golang.org/pkg/crypto/#Signer) interface 

You could regroup these 3 cases together for even less code ! 🙂:

```suggestion
	case crypto.Signer:
```

Reply:
Done.